### PR TITLE
Ream-settings-> User profile -> Create attribute

### PR DESCRIPTION
**Pull request by agagancarczyk**
## Motivation
Create attribute as per designs https://marvelapp.com/prototype/5eed532/screen/82246809

## Brief Description
Most of the functionality is implemented, but due to this PR getting large I have opened a couple of issues for the outstanding work:

- Permission section - checkboxes don't return 'admin' and 'user' when selected - these values are expected by the BE and also additional logic is required for this design https://marvelapp.com/prototype/5eed532/screen/82246809 #2266

- Validations section - added validations not being detected as form values and cannot be sent to the BE #2267
- Validations section - deleting validations not implemented yet due to the above

- General Settings section - as per the old console, setting the "Required" toggle to false should hide the "Required For" and the "Required When" fields #2269

- Editing created attribute not implemented yet #2268

## Verification Steps
1. Go to realm-settings, general tab and select User Profile Enabled
2. Select the User Profile tab and click 'create attribute'
3. Verify that the page is as per designs, complete the form and click save. You should still be able to save the new attribute despite the outstanding work.

## Checklist:

- [ ] Code has been tested locally by PR requester
- [ ] User-visible strings are using the react-i18next framework (useTranslation)
- [ ] Help has been implemented
- [ ] axe report has been run and resulting a11y issues have been resolved
- [ ] Unit tests have been created/updated

### CODE REVIEW

File: src/realm-settings/NewAttributeSettings.tsx [hunk 1]
```diff
@@ -1,23 +1,205 @@
...
+          t("annotations"),
+        ]}
       >
-        <p>{t("createAttribute")}</p>
-      </FormAccess>
-    </PageSection>
+        <AttributeGeneralSettings form={form} />
+        <AttributePermission form={form} />
+        <AttributeValidations />
+        <AttributeAnnotations form={form} />
```
Comment (edewit) [inline src/realm-settings/NewAttributeSettings.tsx:82]: Would be better to use a FormProvider here:
```jsx
<FormProvider {...form}>
//...
```
then you can get the form in the sub components:
```typescript
const form = useFormContext();
```

File: src/realm-settings/NewAttributeSettings.tsx [hunk 2]
```diff
@@ -1,23 +1,205 @@
...
+            variant="primary"
+            type="submit"
+            data-testid="attribute-create"
+          >
+            {t("common:create")}
+          </Button>
+          <Button
+            variant="link"
+            onClick={() =>
+              history.push(toUserProfile({ realm, tab: "attributes" }))
```
Comment (edewit) [inline src/realm-settings/NewAttributeSettings.tsx:96]: you can convert this to a `<Link` e.g.:

```jsx
<Link to={ttoUserProfile({ realm, tab: "attributes" })}>
```

File: src/realm-settings/NewAttributeSettings.tsx [hunk 3]
```diff
@@ -1,23 +1,205 @@
...
+    []
+  );
+
+  useFetch(
+    () => adminClient.clientScopes.find(),
+    (clientScopes) => {
+      setClientScopes(clientScopes);
+    },
+    []
+  );
```
Comment (edewit) [inline src/realm-settings/NewAttributeSettings.tsx:131]: you can combine these two:
```suggestion
  useFetch(
    () =>
      Promise.all([
        adminClient.users.getProfile({ realm: realmName }),
        adminClient.clientScopes.find(),
      ]),
    ([config, clientScopes]) => {
      setConfig(config);
      setClientScopes(clientScopes);
    },
    []
  );
```

File: src/realm-settings/NewAttributeSettings.tsx [hunk 4]
```diff
@@ -1,23 +1,205 @@
...
+
+    const newAttribute = [
+      {
+        name: profileConfig.name,
+        displayName: profileConfig.displayName,
+        required: required,
+        validations: validations,
+        selector: selector,
+        permissions: permissions,
+        annotations: annotations,
```
Comment (edewit) [inline src/realm-settings/NewAttributeSettings.tsx:169]: Can use the short hand like this:
```suggestion
        required,
        validations,
        selector,
        permissions,
        annotations,
```

File: src/realm-settings/user-profile/attribute/AddValidatorDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          toggleDialog={toggleModal}
+          selected={selectedValidator!}
+        />
+      )}
+      <Modal
+        variant={ModalVariant.small}
+        title={t("addValidator")}
+        isOpen
+        onClose={toggleDialog}
+        width={"40%"}
```
Comment (edewit) [inline src/realm-settings/user-profile/attribute/AddValidatorDialog.tsx:47]: is this really needed you already use variant = small

**Comment by edewit** [comment]
There are a couple of controls that are missing a defaultValue see the console log

File: src/realm-settings/user-profile/attribute/AddValidatorDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  toggleDialog,
+  onConfirm,
+}: AddValidatorDialogProps) => {
+  const { t } = useTranslation("realm-settings");
+  const [selectedValidator, setSelectedValidator] = useState<Validator>();
+  const [addValidatorRoleModalOpen, setAddValidatorRoleModalOpen] =
+    useState(false);
+  const toggleModal = () => {
+    setAddValidatorRoleModalOpen(!addValidatorRoleModalOpen);
+  };
```
Comment (edewit) [inline src/realm-settings/user-profile/attribute/AddValidatorDialog.tsx:30]:
```suggestion
  const [addValidatorRoleModalOpen, toggleModal] =
    useToggle();
  };
```

**Comment by agagancarczyk** [comment]
hi @edewit Many thanks for your great feedback. I have now addressed all of the items. Could you please review this PR again for me? Thank you!

File: src/realm-settings/user-profile/attribute/AttributeGeneralSettings.tsx [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+              value: true,
+              message: `${t("validateName")}`,
+            },
+          })}
+          data-testid="attribute-name"
+        />
+        {form.errors.name && (
+          <div className="error">{form.errors.name.message}</div>
+        )}
+      </FormGroup>
```
Comment (edewit) [inline src/realm-settings/user-profile/attribute/AttributeGeneralSettings.tsx:78]: You can simplify this a bit like this:
```suggestion
      <FormGroup
        label={t("attributeName")}
        labelIcon={
          <HelpItem
            helpText="realm-settings-help:attributeNameHelp"
            fieldLabelId="realm-settings:attributeName"
          />
        }
        fieldId="kc-attribute-name"
        isRequired
        helperTextInvalid={t("common:required")}
        validated={
          form.errors.name ? "error" : "default"
        }
      >
        <TextInput
          type="text"
          id="kc-attribute-name"
          name="name"
          defaultValue=""
          ref={register({ required: true })}
          data-testid="attribute-name"
          validated={
             form.errors.name ? "error" : "default"
           }
        />
      </FormGroup>
```

**Review by edewit — APPROVED** [review]

**Pull request merged**